lib/core/src/Cardano/Wallet.hs

```diff
@@ -692,34 +696,40 @@ newWalletLayer tracer bp db nw tl = do
                         then Ready
                         else Restoring progress'
 
-            let txs_cps = applyBlocks @s @t (NE.toList blocks) cp
-            let txs = fold $ fst <$> txs_cps
-
-            let cpLast = fromMaybe cp $ lastMay $ snd <$> txs_cps
-            let (Quantity bhLast) = blockHeight cpLast
+            let putCheckpoint cp txs = do
+                    let meta' = calculateMetadata (view #slotId $ currentTip cp)
+                    DB.putCheckpoint db (PrimaryKey wid) cp
+                    DB.putTxHistory db (PrimaryKey wid) txs
+                    DB.putWalletMeta db (PrimaryKey wid) meta'
+
+            -- We only need to create checkpoints for blocks that are
+            -- "unstable".
+            -- We define unstable blocks as blocks that `k` blocks from the
+            -- network tip, where k is the epoch stability parameter.
+            let (Quantity k) = epochStability
+            let bhUnstable = fromIntegral (flatSlot epochLength nodeTip) - k
```

Comment:
🤔 This appears to be mixing up slots and blocks. I think we should compare against the blockHeight of the tip minus k. 

(See description of #640)

Reply:
Good point!

Reply:
Fixed!